Mancala/src/MancalaBackend/MancalaBoard.java

```diff
@@ -42,6 +46,99 @@ public void startNewGame(PitSize size) {
     public PlayerTypes getTurn() {
         return turn;
     }
+    
+    /**
+     * Get the player A's pit list.
+     * @return player A pits
+     */
+    public MancalaPit[] getPlayerAPits() {
+    	return playerAPits;
+    }
+    
+    /**
+     * Get the player B's pit list.
+     * @return player B pits
+     */
+    public MancalaPit[] getPlayerBPits() {
+    	return playerBPits;
+    }
+    
+    /**
+     * Switch the turn.
+     */
+    public void nextTurn() {
+    	turn = turn == PlayerTypes.PlayerA ? PlayerTypes.PlayerB : PlayerTypes.PlayerA;
+    }
+    
+    /**
+     * The player who has turn pick up stones from the certain pit and placing them.
+     * @param targetPit - the index of pit that the player chose
+     * @return True if the player gains free turn, false if not.
+     */
+    public boolean pickUpStones(int targetPit) {
+    	boolean freeTurn = false;
+    	MancalaPit[] playerPit = turn == PlayerTypes.PlayerA ? playerAPits : playerBPits;
+    	MancalaPit[] opponentPit = turn == PlayerTypes.PlayerA ? playerBPits : playerAPits;
+    	MancalaPit[] currentPit = playerPit;
+    	
+    	//Pull the stones
+    	MancalaStoneCluster cluster = playerPit[targetPit].pullStones();
+    	
+    	//Place the stones
+    	boolean isPlayerPit = true;
+    	boolean passed = true;
+    	
+    	while(passed) {
+    		targetPit++;
+    		//Skip the opponent mancala
+    		if(!isPlayerPit && targetPit == BOARD_SIZE-1) {
+    			targetPit++;
+    		}
+    		
+    		//Switch the side
+    		if(targetPit >= BOARD_SIZE) {
+    			targetPit = 0;
+    			isPlayerPit = !isPlayerPit;
+    			currentPit = isPlayerPit ? playerPit : opponentPit;
+    		}
+    		
+    		//Place a stones
+    		currentPit[targetPit].passStones(cluster);
+    		passed = cluster.getStoneCount() <= 0 ? false : true;
+    		
+    		//Check if the last stone was placing on the empty pit or the mancala.    		
+    		if(cluster.getStoneCount() == 0) {
+    			if(currentPit[targetPit].getContainedCluster().getStoneCount() == 1) {
+        			playerPit[BOARD_SIZE-1].addStones(opponentPit[BOARD_SIZE-2-targetPit].pullStones());
+    				playerPit[BOARD_SIZE-1].addStones(currentPit[targetPit].pullStones());
+        		}
+    			if(currentPit[targetPit].isMancala()) {
+    				freeTurn = true;
+    			}
+    		}
+    	}
+    	return freeTurn;
+    }
+    
+    /**
+     * Check the game is over or not. Game is over when the stones in pits are gone.
+     * @return True if game is over, false if not.
+     */
+    public boolean isGameOver() {
```

Comment:
I feel like we shouldn't require the frontend to have to "check" for a game over, but rather have the frontend be notified of a game over